src/main/java/com/synopsys/integration/detect/workflow/componentlocationanalysis/BdioToComponentListTransformer.java

```diff
@@ -1,15 +1,95 @@
+import com.synopsys.integration.bdio.model.externalid.ExternalId;
 import com.synopsys.integration.detect.workflow.bdio.BdioResult;
+import com.synopsys.integration.detect.workflow.codelocation.DetectCodeLocation;
+import com.synopsys.integration.fixpr.generic.beans.Component;
+
+import java.util.*;
+import java.util.stream.Collectors;
 
 /**
  * Transforms {@link BdioResult} to list of {@link Component}s, which will then be used to assemble the input to
  * Component Locator.
  *
- * Will be fully implemented in a subsequent pull request to the Fix PR feature branch.
  */
 public class BdioToComponentListTransformer {
-    public ComponentLocatorInput transformBdioToComponentList(BdioResult bdio) {
-        return new ComponentLocatorInput(null, null, null);
+    /**
+     * Given a BDIO, creates a list containing each detected component's corresponding {@link Component} representation.
+     * Each component is included once, duplicates are ignored.
+     * @param bdio
+     * @return list of unique {@link Component}s
+     */
+    public List<Component> transformBdioToComponentList(BdioResult bdio) {
+        List<ExternalId> allExternalIds = new ArrayList<>();
+
+        Set<DetectCodeLocation> codeLocations = bdio.getCodeLocationNamesResult().getCodeLocationNames().keySet();
+        for (DetectCodeLocation cl : codeLocations) {
+            List<ExternalId> allDepsForThisCodeLocation = processDependencyGraph(cl.getDependencyGraph(), cl.getDependencyGraph().getDirectDependencies(), new HashSet<>());
+            allExternalIds.addAll(allDepsForThisCodeLocation);
+        }
+
+        return externalIDsToComponentList(allExternalIds);
+    }
+
+    /**
+     * Starting with a list of the direct dependencies of a code location, recursively collects {@link ExternalId}s of
+     * all direct and transitive dependencies.
+     * @param graph
+     * @param dependencies
+     * @param alreadyProcessedDependencies
+     * @return list of all dependencies in the original dependency graph
+     */
+    private List<ExternalId> processDependencyGraph(
+            DependencyGraph graph,
+            Set<Dependency> dependencies,
+            HashSet<ExternalId> alreadyProcessedDependencies
+    ) {
+        List<ExternalId> addedDependencies = new ArrayList<>();
+        for (Dependency dependency : dependencies) {
+            if (!alreadyProcessedDependencies.contains(dependency.getExternalId())) {
+                addedDependencies.add(dependency.getExternalId());
+                alreadyProcessedDependencies.add(dependency.getExternalId());
+                Set<Dependency> transitiveDependencies = graph.getChildrenForParent(dependency);
+                List<ExternalId> addedChildren = processDependencyGraph(graph, transitiveDependencies, alreadyProcessedDependencies);
+                addedDependencies.addAll(addedChildren);
+            }
+        }
+        return addedDependencies;
+    }
```

Comment:
I have seen this recursion approach before where a new List is created in every recursive call i.e. `List<ExternalId> addedDependencies = new ArrayList<>();` 
I think it is acceptable in situations where the dependency tree depth is not too high.
To be cautious, in this case, you could switch the return type to void and optimize the method by having a class attribute for `addedDependencies` and have this recursive method fill up the ArrayList `addedDependencies`. This will avoid creating `new` ArrayLists at each recursion level that is temporarily being saved on the implicit stack. You could also pass by ref the `addedDependencies` array list if you want to avoid a class attribute for some reason.

Reply:
Will do, please see IDETECT-3950